R-package/R/lgb.train.R

```diff
@@ -270,11 +270,11 @@ lgb.train <- function(params = list(),
   # When early stopping is not activated, we compute the best iteration / score ourselves by selecting the first metric and the first dataset
   if (record && length(valids) > 0 && is.na(env$best_score)) {
     if (env$eval_list[[1]]$higher_better[1] == TRUE) {
-      booster$best_iter <- unname(which.max(unlist(booster$record_evals[[2]][[1]][[1]])))
-      booster$best_score <- booster$record_evals[[2]][[1]][[1]][[booster$best_iter]]
+      booster$best_iter <- unname(which.max(unlist(booster$record_evals[[names(valids)[1]]][[1]][[1]])))
+      booster$best_score <- booster$record_evals[[names(valids)[1]]][[1]][[1]][[booster$best_iter]]
     } else {
-      booster$best_iter <- unname(which.min(unlist(booster$record_evals[[2]][[1]][[1]])))
-      booster$best_score <- booster$record_evals[[2]][[1]][[1]][[booster$best_iter]]
+      booster$best_iter <- unname(which.min(unlist(booster$record_evals[[names(valids)[1]]][[1]][[1]])))
+      booster$best_score <- booster$record_evals[[names(valids)[1]]][[1]][[1]][[booster$best_iter]]
```

Comment:
I think we need to fix this in `lgb.cv.R` too, right?

https://github.com/microsoft/LightGBM/blob/master/R-package/R/lgb.cv.R#L309

Since for now this logic is duplicated in both places.

Reply:
also would it be difficult to add two unit tests (one for either side of the `else`) to ensure we don't re-introduce this bug in the future?

If it would be, then could you open an issue documenting the need for such tests and attach it to #2302 ?